File: Data Structures & Algorithms/two-integer-sum-ii/submission-3.java
class Solution {
    public int total(int n1, int n2){
        return n1 + n2;
    }
    public int[] twoSum(int[] nums, int target) {
        int l = 0, r = nums.length - 1;

        while(l < r){
            if(total(nums[l], nums[r]) < target)
                l++;
            else if(total(nums[l], nums[r]) > target)
                r--;
            else
                return new int[] {l + 1, r + 1};
        }

        return new int[] {};
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        total = sum(nums)

        if total % 2:
            return False

        target = total // 2

        cache = defaultdict(bool)

        def helper(i, target):
            if target == 0:
                return True
            if i == len(nums):
                return False

            if (i, target) in cache:
                return cache[(i, target)]

            res = helper(i + 1, target)

            if nums[i] <= target:
                res |= helper(i + 1, target - nums[i])

            cache[(i, target)] = res

            return res

        return helper(0, target)

